File: src/tools/run-command.test.ts
import assert from "node:assert/strict";
import { mkdtemp, rm } from "node:fs/promises";
import os from "node:os";
import path from "node:path";
import test from "node:test";

import { Workspace } from "../workspace/workspace.js";
import { ToolRegistry } from "./registry.js";
import { RunCommandTool } from "./run-command.js";
import { AllowAllApprover, type ToolResult } from "./types.js";

const signal = new AbortController().signal;

async function commandTool(
  t: test.TestContext,
): Promise<{ readonly root: string; readonly run: (value: object) => Promise<ToolResult> }> {
  const root = await mkdtemp(path.join(os.tmpdir(), "nash-command-"));
  t.after(async () => {
    await rm(root, { recursive: true, force: true });
  });
  const workspace = await Workspace.open(root);
  const tool = new RunCommandTool(workspace);
  const registry = new ToolRegistry([tool]);
  return {
    root: workspace.root,
    run: async (value) =>
      await registry.execute(
        {
          id: "command-1",
          name: "run_command",
          arguments: JSON.stringify(value),
        },
        new AllowAllApprover(),
        signal,
      ),
  };
}

function kindOf(result: ToolResult): unknown {
  return result.metadata?.kind;
}

test("RunCommandTool uses the workspace and reports nonzero exits as results", async (t) => {
  const { root, run } = await commandTool(t);

  const cwd = await run({ command: "pwd" });
  assert.equal(cwd.isError, false);
  assert.match(cwd.content, new RegExp(escapeRegExp(root)));

  const failed = await run({ command: "printf boom; exit 7" });
  assert.equal(failed.isError, true);
  assert.equal(kindOf(failed), "command_failed");
  assert.equal(failed.metadata?.exit_code, 7);
  assert.match(failed.content, /boom/);
});

test("RunCommandTool terminates a command at its timeout", async (t) => {
  const { run } = await commandTool(t);

  const result = await run({ command: "sleep 5", timeout_seconds: 1 });

  assert.equal(result.isError, true);
  assert.equal(kindOf(result), "timeout");
});

test("RunCommandTool strips sensitive environment variables", async (t) => {
  const previous = {
    apiKey: process.env.NASH_TEST_API_KEY,
    token: process.env.NASH_TEST_TOKEN,
    safe: process.env.NASH_TEST_SAFE,
  };
  process.env.NASH_TEST_API_KEY = "do-not-leak";
  process.env.NASH_TEST_TOKEN = "do-not-leak-either";
  process.env.NASH_TEST_SAFE = "visible";
  t.after(() => {
    restoreEnvironment("NASH_TEST_API_KEY", previous.apiKey);
    restoreEnvironment("NASH_TEST_TOKEN", previous.token);
    restoreEnvironment("NASH_TEST_SAFE", previous.safe);
  });
  const { run } = await commandTool(t);
  const command =
    `printf '%s|%s|%s|%s' ` +
    `"\${NASH_TEST_API_KEY-unset}" "\${NASH_TEST_TOKEN-unset}" ` +
    `"\${NASH_TEST_SAFE-unset}" "\${NASH_AGENT-unset}"`;

  const result = await run({ command });

  assert.equal(result.isError, false);
  assert.match(result.content, /unset\|unset\|visible\|1/);
  assert.doesNotMatch(result.content, /do-not-leak/);
});

test("RunCommandTool keeps bounded head and tail output", async (t) => {
  const { run } = await commandTool(t);
  const command = `awk 'BEGIN { for (i=0; i<70000; i++) printf "x" }'`;

  const result = await run({ command });

  assert.equal(result.isError, false);
  assert.equal(result.metadata?.truncated, true);
  assert.equal(result.metadata?.output_bytes, 70_000);
  assert.match(result.content, /bytes omitted/);
});

function restoreEnvironment(key: string, value: string | undefined): void {
  if (value === undefined) {
    delete process.env[key];
  } else {
    process.env[key] = value;
  }
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

File: internal/tool/run_command_test.go
package tool

import (
	"context"
	"fmt"
	"strings"
	"testing"
)

func TestRunCommandUsesWorkspaceAndReportsNonzeroExit(t *testing.T) {
	current, root := newTestWorkspace(t)
	runner := NewRunCommand(current)

	cwd := runner.Execute(context.Background(), `{"command":"pwd"}`)
	if cwd.IsError {
		t.Fatalf("pwd failed: %#v", cwd)
	}
	if !strings.Contains(cwd.Content, root) {
		t.Fatalf("command output %q does not contain workspace %q", cwd.Content, root)
	}

	failed := runner.Execute(context.Background(), `{"command":"printf boom; exit 7"}`)
	if !failed.IsError {
		t.Fatalf("expected nonzero exit result, got %#v", failed)
	}
	if got, ok := failed.Metadata["exit_code"].(int); !ok || got != 7 {
		t.Fatalf("exit code = %#v, want 7", failed.Metadata["exit_code"])
	}
	if got := resultKind(t, failed); got != "command_failed" {
		t.Fatalf("failure kind = %q, want command_failed", got)
	}
	if !strings.Contains(failed.Content, "boom") {
		t.Fatalf("missing command output: %q", failed.Content)
	}
}

func TestRunCommandTimesOut(t *testing.T) {
	current, _ := newTestWorkspace(t)
	runner := NewRunCommand(current)

	result := runner.Execute(context.Background(), `{"command":"sleep 2","timeout_seconds":1}`)
	if !result.IsError || resultKind(t, result) != "timeout" {
		t.Fatalf("expected timeout result, got %#v", result)
	}
}

func TestRunCommandStripsSensitiveEnvironment(t *testing.T) {
	t.Setenv("NASH_TEST_API_KEY", "do-not-leak")
	t.Setenv("NASH_TEST_TOKEN", "do-not-leak-either")
	t.Setenv("NASH_SAFE_VALUE", "visible")
	current, _ := newTestWorkspace(t)
	runner := NewRunCommand(current)

	command := `printf '%s|%s|%s|%s' "${NASH_TEST_API_KEY-unset}" "${NASH_TEST_TOKEN-unset}" "${NASH_SAFE_VALUE-unset}" "${NASH_AGENT-unset}"`
	result := runner.Execute(context.Background(), fmt.Sprintf(`{"command":%q}`, command))
	if result.IsError {
		t.Fatalf("environment probe failed: %#v", result)
	}
	if !strings.Contains(result.Content, "unset|unset|visible|1") {
		t.Fatalf("unexpected sanitized environment: %q", result.Content)
	}
	if strings.Contains(result.Content, "do-not-leak") {
		t.Fatalf("sensitive value leaked in output: %q", result.Content)
	}
}

func TestRunCommandTruncatesOutput(t *testing.T) {
	current, _ := newTestWorkspace(t)
	runner := NewRunCommand(current)

	command := `awk 'BEGIN { for (i=0; i<70000; i++) printf "x" }'`
	result := runner.Execute(context.Background(), fmt.Sprintf(`{"command":%q}`, command))
	if result.IsError {
		t.Fatalf("large output command failed: %#v", result)
	}
	if truncated, ok := result.Metadata["truncated"].(bool); !ok || !truncated {
		t.Fatalf("expected truncation metadata, got %#v", result.Metadata)
	}
	if bytes, ok := result.Metadata["output_bytes"].(int64); !ok || bytes != 70000 {
		t.Fatalf("output bytes = %#v, want 70000", result.Metadata["output_bytes"])
	}
	if !strings.Contains(result.Content, "bytes omitted") {
		t.Fatalf("missing truncation marker in output")
	}
}
